### ISSUE TITLE
Basic question finally switched as multiple probe checking/connecting

### ISSUE THREAD

**Issue by alkabal** [linuxcnc-probe-screen/probe-screen-ng#12]
Hi

I'm sorry to ask here, if you are not agree you can close this issue, but maybe this is interesting to add this feature for all

Can i have your opinion : 
https://forum.linuxcnc.org/10-advanced-configuration/40817-3d-touch-probe-tool-setter-z-touch#191458

Br

**Comment by alkabal** [comment]
Other question is 

did you understand why in some macro they use : 
G38.2 Z[-#<_hal[probe.ps_probe_max]>]
and other use : 
G38.2 Z #<_ini[TOOLSENSOR]MAXPROBE>

**Comment by alkabal** [comment]
Link for full guide that can be usefull to do something similar here
https://vers.by/en/blog/useful-articles/probe-screen

**Comment by alkabal** [comment]
About feature : Settings for the tool diameter measurement button

They write : 
makes a measurement of the height of the toolsetter's pad, then starts the spindle at reverse(it is recommended to set no more than 1500 rpm) and measures the diameter by analogy with the center of the cylinder.

But i see nothing about that in the code, i think this need to be added with deactivable option for people with only Z tool setter and maybe add a pause and ask user confirmation before starting spindle.

**Comment by freisei** [comment]
> Other question is
> 
> did you understand why in some macro they use :
> G38.2 Z[-#<_hal[probe.ps_probe_max]>]
> and other use :
> G38.2 Z #<_ini[TOOLSENSOR]MAXPROBE>

The Z[-#<_hal[probe.ps_probe_max]>] is a Value from probe-screen-GUI "Max probe distance", the other is from Ini for searching the toolsetter-pad.

**Comment by alkabal** [comment]
Thanks for this, you are right (i appologize for this stupid question yesterday i was too much sleepless)

Most important is are you agree for add starting spindle reverse management for tool diameter measurement ?
And about missing probe or tool setter detection, it is something that can be added here in you opinion ?

Br

**Comment by freisei** [comment]
> Most important is are you agree for add starting spindle reverse management for tool diameter measurement ?

Here is the code doing tool diameter measurement:
https://github.com/linuxcnc-probe-screen/probe-screen-ng/blob/18873ce9fe9c71221fb1b5ffe7d40d37a839acce/psng/psng.py#L2384

Of course you can implement a macro for starting and stopping spindle there.

>  And about missing probe or tool setter detection, it is something that can be added here in you opinion ?
What do you mean with that?

**Comment by kiall** [comment]
> Most important is are you agree for add starting spindle reverse management for tool diameter measurement ?

Personally, I wouldn't want my spindle moving during tool measurement. It looks like vers.by never implemented this either, and their docs are just incorrect. Is this really necessary to add?

> And about missing probe or tool setter detection, it is something that can be added here in you opinion ?

Is there a way to reliably detect when the touch probe or tool setter isn't connected? The only way I can think of is to have the probe and tool setter as NC, that way - if the voltage is not present, you can be sure the probe isn't connected. For NC probes, G38.n already detects this issue and aborts - so there's nothing to add.

For NO probes, there is no way to distinguish "probe is not connected" from "probe isn't touching anything" - so I'm not sure there is anything that can be added?

> BTW: you should add an issue for each issue. And not adding more issues to the pending ones. Even the title of this issue is not good at all.

:+1:

**Comment by alkabal** [comment]
> > Most important is are you agree for add starting spindle reverse management for tool diameter measurement ?
> 
> Personally, I wouldn't want my spindle moving during tool measurement. It looks like vers.by never implemented this either, and their docs are just incorrect. Is this really necessary to add?
> 
> > And about missing probe or tool setter detection, it is something that can be added here in you opinion ?
> 
> Is there a way to reliably detect when the touch probe or tool setter isn't connected? The only way I can think of is to have the probe and tool setter as NC, that way - if the voltage is not present, you can be sure the probe isn't connected. For NC probes, G38.n already detects this issue and aborts - so there's nothing to add.
> 
> For NO probes, there is no way to distinguish "probe is not connected" from "probe isn't touching anything" - so I'm not sure there is anything that can be added?
> 
> > BTW: you should add an issue for each issue. And not adding more issues to the pending ones. Even the title of

**Comment by alkabal** [comment]
> > Most important is are you agree for add starting spindle reverse management for tool diameter measurement ?
> 
> Here is the code doing tool diameter measurement:
> https://github.com/linuxcnc-probe-screen/probe-screen-ng/blob/18873ce9fe9c71221fb1b5ffe7d40d37a839acce/psng/psng.py#L2384
> 
> Of course you can implement a macro for starting and stopping spindle there.

Thanks for pointing to the right place, this is something i like to experiment, the vers TSM allow to measure in xyz direction so this is possible to be interesting for create tool table.


> Personally, I wouldn't want my spindle moving during tool measurement. It looks like vers.by never implemented this either, and their docs are just incorrect. Is this really necessary to add?

Maybe this can added as a configurable option with some warning and M0 for prevent unwanted things.


Off course i can check that for myself i only like to speak about that with you.

Thanks

**Comment by freisei** [comment]
> For NO probes, there is no way to distinguish "probe is not connected" from "probe isn't touching anything" - so I'm not sure there is anything that can be added?

this can be realized with a additional input only for checking.
You need a high-ohm resistor (~ 10K)  in the TS or probe and another resistor (~1K) on your input. 
You can find much infos about that (in german: Kabelbruchüberwachung)

**Comment by alkabal** [comment]
> Is there a way to reliably detect when the touch probe or tool setter isn't connected? The only way I can think of is to have the probe and tool setter as NC, that way - if the voltage is not present, you can be sure the probe isn't connected. For NC probes, G38.n already detects this issue and aborts - so there's nothing to add.
> 
> For NO probes, there is no way to distinguish "probe is not connected" from "probe isn't touching anything" - so I'm not sure there is anything that can be added?
> 

https://forum.linuxcnc.org/10-advanced-configuration/40817-3d-touch-probe-tool-setter-z-touch#191458
For NO NPN this is possible to check if connected or not

For the moment this thread is only for clarification, but yes if you agree for implementing something like this i can open separate thread, i hope this is ok for you or

**Comment by alkabal** [comment]
using two input from card and using hal var in the ocode allow to check both sensor checking

I have really not think about other type of probe but i think this is possible for other situation ???

**Comment by freisei** [comment]
in general i think this isnt very relevant to PSNG. All that can be done in HAL.
The only thing in my optinion that could be done good in PSNG would be a "Test-Press". That means the probe or TS must be triggered by hand one time before PSNG starts moving down.

**Comment by alkabal** [comment]
Insoide the icin foler one is already made for this

**Comment by alkabal** [comment]
about the psng unraleted stuff, i can understand off course but i does not see how to do this for the moment without implementing a hal input pin check inside macro.

**Comment by alkabal** [comment]
> Inside the icon foler one is already made for this purpose (hand touch)

maybe this is enough to force user to test 3D probe at first use.

**Comment by kiall** [comment]
> > Is there a way to reliably detect when the touch probe or tool setter isn't connected? The only way I can think of is to have the probe and tool setter as NC, that way - if the voltage is not present, you can be sure the probe isn't connected. For NC probes, G38.n already detects this issue and aborts - so there's nothing to add.
> > For NO probes, there is no way to distinguish "probe is not connected" from "probe isn't touching anything" - so I'm not sure there is anything that can be added?
> 
> https://forum.linuxcnc.org/10-advanced-configuration/40817-3d-touch-probe-tool-setter-z-touch#191458
> For NO NPN this is possible to check if connected or not with pull up
> Pin is low if nothing is connected (generating error if i try to use probe)
> Pin is High if TSM + converter is connected or if touch plate + pullup
> Pin is low if probe is triggered

What you've done is effectively changed the probe from a NO to an NC probe. G38.x will abort if you don't connect your probe (assuming the electronics all work as described).

That said, because the entirety of the probe circuit isn't actually setup as NC, you're not necessarily getting the same safety guarantees you would normally get from a true NC probe. There is still a portion of your probe circuit which is NO, and anything that breaks / is disconnected in that part is undetectable.

**Comment by alkabal** [comment]
Are you sure ? 

Disconnecting wire from the TSM before converter and pin get low (because TSM in NC the converter can see the input change)
Disconnecting converter from mesa get pin low
Put converter off with integrated button get pin low
Pin is high only when converter is correctly connected to mesa and TSM is connected to converter
At the end (for what i think in my situation) not connecting probe is similar to probe already triggered and does not allow G38 execution so hope this is enough for connecting break detection. (i see only a issue if input probe wire is short circuited
with VCC)
Touch plate with deported pull-up connected like my picture on the forum work same way

For now i haven't the 3D probe but i hope with same setting NO NPN this is similar.

**Comment by kiall** [comment]
> Touch plate with deported pull-up connected like my picture on the forum work same way

Cut the ground wire running to the tool in your picture, or just don't connect it, and the error will not be detected as it is in the end a NO system :)

There's no diagram of how your wiring up the TSM, but it's a NC device - so you would be far better off wiring this directly in as is / without modification and using HAL to combine the touch plate and tool setter inputs.

**Comment by alkabal** [comment]
I have no ground wire on the tool, but you are right machine is grounded to mesa card. So yes here is a possible issue, thanks a lot for pointing this !

Oh yes you are 100% right if I use separate input pin + Hal stuff this is really better like this !

I have for now think about that with only using one input and get compatible with 3 device but this is not the really the wanted result so you are really right about device selection.

Many thanks for all this unrelated information !!!

**Comment by alkabal** [comment]
Now the next things is continue discussion for checking probe and tool setter connection inside macro ? Inside python ? Or you see other way for this directly in Hal  ? 

And the last things is about including a stuff like this in the code (or keep this only for my own usage)
If this is only for my own I think a touch test before first probe request as discussed can be great.

Br

**Comment by kiall** [comment]
> Now the next things is continue discussion for checking probe and tool setter connection inside macro ? Inside python ? Or you see other way for this directly in Hal ?

I'm not sure there is anything to implement in the probe screen for this. It's all very dependant on the probe(s) someones has, how they choose to wire them, what interface card they have etc.

LinuxCNC has a single pin (`motion.probe-input`) where it expects a NC input to be connected. Everything that leads from the physical probe(s) to that PIN can be different for each machine and needs to done by the user in HAL.

Since everything prior to `motion.probe-input` is unique to the specific machine, and given G38.x already verifies that `motion.probe-input` is high prior to beginning a probe move - I don't see what can be added that is both generic (i.e. not specific to your machine) and isn't already done by G38.x.

> And the last things is about including a stuff like this in the code (or keep this only for my own usage)
> If this is only for my own I think a touch test before first probe request as discussed can be great.

Adding the ability to test a probe is functioning makes sense - I currently do this with a simple pyvcp "LED" connected to `motion.probe-input` - I touch the touch probe, the "LED" turns green. I touch the tool setter, it "LED" turns green. Adding an "LED" to the probe panel makes sense and can be done in generic way since it only relies on `motion.probe-input`.

**Comment by alkabal** [comment]
Maybe better to do this in python code

**Comment by alkabal** [comment]
And this only need to be linked to corresponding input pin in the Hal config, user can invert pin or use only one using net

**Comment by kiall** [comment]
> But from HAL i have no idea how to determining if the probe motion is for tool setter or touchprobe, as i understand this is something done in the macro.
> ...
> so adding 2 hal input and check the correspondance in the macro as needed

The macro (or python code) can't tell this in a generic way either - there is only 1 PIN that is always available `motion.probe-input`. If a Macro or Python expects more input signals, then it's no longer generic - e.g. it won't work correctly with my setup where I have two NC probes wired to a single physical input pin (Some of the vers.by probes allow for daisy chaining NC probes/sensors together in this way).

I have an idea on how to allow machine specific stuff like this, without having to add it for everyone. I'll push a PR this evening.

**Comment by alkabal** [comment]
Oh many thanks for interest about this stuff !!

**Comment by kiall** [comment]
> I think this is not a real problem with your setup but you can't make the difference if on or other is missing, but how do you use only the tool setter withour probe ? you keep it connected ?

Yes, both are always wired in and ready to use. Without dropping to a NO probe (which, for safety reasons I absolutely don't want to do), LinuxCNC doesn't offer the flexibility to handle multiple probes without them all being powered on.

--

https://github.com/linuxcnc-probe-screen/probe-screen-ng/pull/15 - Have a look at this PR - it should allow you to add extra machine specific code at the start of any macro, without having to modify the built in ones.

**Comment by alkabal** [comment]
Hi

Because merge is done the topic is close so i come back here.

I have pushed my last modification

I have added some comment i like if you can review.

https://github.com/alkabal/probe-screen-ng

**Comment by alkabal** [comment]
I do some change regarding to saved pref value because there is no reason to lost the setting too much easy as it is actually

What i like to do is increase possibility and safety.

With the revisited version i like to make possible :
-Probing like original
-Toolchange like original
-Allowing possibility to create a tooltable for Z offset if machine changing tool is enough accurate
-Allowing possibility to create a tooltable for Z offset and Diameter (Z can be corrected later each time with autolenght if needed)
-Make things more logic for preparatory function


I have do another preview for GUI modification
![GUI change proposition](https://user-images.githubusercontent.com/33371787/102301982-56bf9480-3f58-11eb-8235-b77eef6174c4.png)


MAYBE I'M COMPLETLY WRONG


Any suggestion about this and also about comment in my fork are really really welcome.

br

**Comment by alkabal** [comment]
![GUI change proposition](https://user-images.githubusercontent.com/33371787/102302603-a94d8080-3f59-11eb-8995-0876f12c3e38.png)


this is more close to what is needed

**Comment by kiall** [comment]
> Because merge is done the topic is close so i come back here.
> 
> I have pushed my last modification
> 
> I have added some comment i like if you can review.
> 
> https://github.com/alkabal/probe-screen-ng

Hey! Honestly, there are so many changes in there - it's really difficult to review in a meaningful way. It looks like there is 10 or more unrelated changes all combined into the diff, reviewing that and ensuring it's all safe+valid is next to impossible.

**Comment by alkabal** [comment]
Hi

Saad but i understand lol, many thanks for trying to check !

Br
